ldap/servers/slapd/auditlog.c

```diff
@@ -266,20 +301,285 @@ add_entry_attrs(Slapi_Entry *entry, lenstr *l)
                 strcasecmp(attr, CONFIG_ROOTPW_ATTRIBUTE) == 0)
             {
                 /* userpassword/rootdn password - mask the value */
-                addlenstr(l, "#");
-                addlenstr(l, attr);
-                addlenstr(l, ": ****************************\n");
+                if (use_json) {
+                    json_object *secret_obj = json_object_new_object();
+                    json_object_object_add(secret_obj, attr,
+                                           json_object_new_string("**********************"));
+                    json_object_array_add(id_list, secret_obj);
+                } else {
+                    addlenstr(l, "#");
+                    addlenstr(l, attr);
+                    addlenstr(l, ": ****************************\n");
+                }
                 continue;
             }
-            log_entry_attr(entry_attr, attr, l);
+            if (use_json) {
+                log_entry_attr_json(entry_attr, attr, id_list);
+            } else {
+                log_entry_attr(entry_attr, attr, l);
+            }
         }
     }
+    if (json_object_array_length(id_list) > 0) {
+        json_object_object_add(json_log, "id_list", id_list);
+    } else {
+        /* free empty list */
+        json_object_put(id_list);
+    }
     slapi_ch_free_string(&display_attrs);
 }
 
+static void
+add_entry_attrs(Slapi_Entry *entry, lenstr *l)
+{
+    add_entry_attrs_ext(entry, l, PR_FALSE, NULL);
+}
+
+static void
+add_entry_attrs_json(Slapi_Entry *entry,  json_object *json_log)
+{
+    add_entry_attrs_ext(entry, NULL, PR_TRUE, json_log);
+}
+
+static void
+write_audit_file_json(Slapi_PBlock *pb, Slapi_Entry *entry, int logtype,
+                      int optype, const char *dn, void *change, int flag,
+                      time_t curtime, int result, int log_format)
+{
+    Connection *pb_conn = NULL;
+    Slapi_Operation *operation = NULL;
+    Slapi_DN *target_sdn = NULL;
+    Slapi_Entry *e = NULL;
+    LDAPMod **mods = NULL;
+    json_object *log_json = NULL;
+    json_object *del_obj = NULL;
+    json_object *modrdn_obj = NULL;
+    struct tm tms;
+    const char *target_dn = NULL;
+    char binddn[JBUFSIZE] = {0};
+    char local_time[JBUFSIZE] = {0};
+    char *client_ip = NULL;
+    char *server_ip = NULL;
+    char *newrdn = NULL;
+    char *mod_dn = NULL;
+    char *msg = NULL;
+    char *time_format = NULL;
+    char *timestr = NULL;
+    char *tmp = NULL;
+    char *tmpsave = NULL;
+    uint64_t conn_id = 0;
+    int32_t rc = -1;
+    int32_t op_id = 0;
+
+    time_format = config_get_auditlog_time_format();
+    (void)localtime_r(&curtime, &tms);
+    if (strftime(local_time, JBUFSIZE, time_format, &tms) == 0) {
```

Comment:
Should we really convert the timestamp in the json version ?
The local time string format is better for the human, but json logs are intended for some post-processing tools 
and IMHO the raw UTC time as integer is usually better for tools (Because each tool may then easily convert it the way it need it.) 
For example the replication latency tool has to convert back the access log timestamp to integer because the different logs may be in different timezone)
And storing directly the integer is faster.

Reply:
Please check this discussion https://github.com/389ds/389ds.github.io/pull/14#discussion_r1575708169
I asked for ISO 8601 Date string because it's unambiguous, human-readable, and well-supported.

Reply:
I agree with ISO 8601 but I do not think we should allow to change the format
 and in any case it should be gmtime and not localtime. 
IMHO allowing to change the format may not be a good idea. 
 Not only it add one more parameter, but there is a big difference between the two formats:
- The ascii output is intended to be directly readable by humans.
- The json output must be post-processed by a tools
And if the timestamp format may change, the tool developper will have to take care of the timestamp format 
If it is only ISO 860, he has only to take care of ISO 8601 format.
So IMHO we should have a single ISO 8601 format and let the tool format the date
Sorry that I have not realized that point in April !

Reply:
@vashirov @progier389 The reason I made the time format configurable is because customers/community have been asking for this for years (for DS and IDM).  I see no harm in having it, but if you guys think it's a bad idea then I will remove it (no problem).

Reply:
> I agree with ISO 8601 but I do not think we should allow to change the format and in any case it should be gmtime and not localtime. 

FYI, the "time" key, which is also the local time.  This is how the current audit log records the time.  So I kept that for legacy purposes and then introduced the "date" key for the custom time.  Again, this is an open discussion but I just wanted clarify what the current PR is doing.

Reply:
You are right: I overlook the **time** key.
 So forgot what i previously wrote: I do not really care about the way the **date** value is formatted ( local time and a configurable format is fine)
 But for "time" you should use ISO 8601 as viktor mentionned. 
  (i.e: using  strftime(isotime, (sizeof isotime), "%FT%TZ", gmtime(&curtime));

rather than format_localTime

Reply:
Oups! should rather use gmtime_r instead of gm_time in multi threaded environment ! :wink: